central/graphql/resolvers/image_components_test.go

```diff
@@ -0,0 +1,434 @@
+	pgtest.CloseGormDB(s.T(), s.gormDB)
+	s.db.Close()
+}
+
+func (s *GraphQLImageComponentTestSuite) TestUnauthorizedImageComponentEndpoint() {
+	_, err := s.resolver.ImageComponent(s.ctx, IDQuery{})
+	s.Error(err, "Unauthorized request got through")
+}
+
+func (s *GraphQLImageComponentTestSuite) TestUnauthorizedImageComponentsEndpoint() {
+	_, err := s.resolver.ImageComponents(s.ctx, PaginatedQuery{})
+	s.Error(err, "Unauthorized request got through")
+}
+
+func (s *GraphQLImageComponentTestSuite) TestUnauthorizedImageComponentCountEndpoint() {
+	_, err := s.resolver.ImageComponentCount(s.ctx, RawQuery{})
+	s.Error(err, "Unauthorized request got through")
+}
+
+func (s *GraphQLImageComponentTestSuite) TestImageComponents() {
+	ctx := SetAuthorizerOverride(s.ctx, allow.Anonymous())
+
+	expectedIDs := []string{"comp1#0.9#", "comp2#1.1#", "comp3#1.0#", "comp4#1.0#"}
+	expectedCount := int32(len(expectedIDs))
+
+	comps, err := s.resolver.ImageComponents(ctx, PaginatedQuery{})
+	s.NoError(err)
+	s.Equal(expectedCount, int32(len(comps)))
+	idList := getIDList(ctx, comps)
+	s.ElementsMatch(expectedIDs, idList)
+
+	count, err := s.resolver.ImageComponentCount(ctx, RawQuery{})
+	s.NoError(err)
+	s.Equal(expectedCount, count)
+}
+
+func (s *GraphQLImageComponentTestSuite) TestImageComponentsScoped() {
+	ctx := SetAuthorizerOverride(s.ctx, allow.Anonymous())
+
+	imageCompTests := []struct {
+		name        string
+		id          string
+		expectedIDs []string
+	}{
+		{
+			"sha1",
+			"sha1",
+			[]string{"comp1#0.9#", "comp2#1.1#", "comp3#1.0#"},
+		},
+		{
+			"sha2",
+			"sha2",
+			[]string{"comp1#0.9#", "comp3#1.0#", "comp4#1.0#"},
+		},
+	}
+
+	for _, test := range imageCompTests {
+		s.T().Run(test.name, func(t *testing.T) {
+			image := s.getImageResolver(ctx, test.id)
+			expectedCount := int32(len(test.expectedIDs))
+
+			comps, err := image.ImageComponents(ctx, PaginatedQuery{})
+			s.NoError(err)
+			s.Equal(expectedCount, int32(len(comps)))
+			idList := getIDList(ctx, comps)
+			s.ElementsMatch(test.expectedIDs, idList)
+
+			count, err := image.ImageComponentCount(ctx, RawQuery{})
+			s.NoError(err)
+			s.Equal(expectedCount, count)
+		})
+	}
+}
+
+func (s *GraphQLImageComponentTestSuite) TestImageComponentMiss() {
+	ctx := SetAuthorizerOverride(s.ctx, allow.Anonymous())
+
+	compID := graphql.ID("invalid")
+
+	_, err := s.resolver.ImageComponent(ctx, IDQuery{ID: &compID})
+	s.Error(err)
+}
+
+func (s *GraphQLImageComponentTestSuite) TestImageComponentHit() {
+	ctx := SetAuthorizerOverride(s.ctx, allow.Anonymous())
+
+	compID := graphql.ID("comp1#0.9#")
+
+	comp, err := s.resolver.ImageComponent(ctx, IDQuery{ID: &compID})
+	s.NoError(err)
+	s.Equal(compID, comp.Id(ctx))
+}
+
+func (s *GraphQLImageComponentTestSuite) TestImageComponentImages() {
+	ctx := SetAuthorizerOverride(s.ctx, allow.Anonymous())
+
+	imageCompTests := []struct {
+		name        string
+		id          string
+		expectedIDs []string
+	}{
+		{
```

Comment:
would be good to test with two images with same components but different operating systems:
image.id = <<some_sha>>
image.scan.operatingSystem = <<some_os>>

image.id = <<some_other_sha>>
image.scan.operatingSystem = <<some_other_os>>